Fixes #3338. `Application.Run/End` -> Callers must dispose Toplevel

tig opened the pull request:
Caused by #3300

## Fixes

- Fixes #3338 

## Proposed Changes/Todos

- [x] Find and fix all occurances of code not cleaning up the Toplevel used for `Run`

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

tig commented:
@BDisp Highly likely I messed up the BackgroundWorker scenarios here. They are convoluted examples and very hard to understand. I did my best and I think they are fine, but since you wrote them you should probably take a close look.

BDisp commented:
The `BackgroundWorker` scenario is really broken. Here is failing because the `Current` toplevel is already disposed and `Running` is still true and responding to mouse report position. I'm afraid something is still not well yet.

![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/c5e97cd6-012a-4be5-af4c-fb8150d0bc6f)

BDisp commented on UnitTests/Dialogs/DialogTests.cs:
```diff
@@ -1331,4 +1331,33 @@ public void Zero_Buttons_Works ()
+    [Fact]
+    [SetupFakeDriver]
+    [TestRespondersDisposed]
+    public void Run_Does_Not_Dispose_Dialog ()
```
Did you realized that running this unit test alone never ends because it's using `WindowsDriver` and not `FakeDriver`. Also the `Ready` event isn't invoked because it's called after the `if (MainLoop.Running && MainLoop.EventsPending ())` where it never get out.

BDisp reviewed (changes requested):
I'm still think we don't need dispose all the views because the `Application` already handle it in the `End` and in the `Shutdown`,

BDisp commented:
I sent the PR https://github.com/tig/Terminal.Gui/pull/24 to this branch. I hope you like it.

tig commented:
I addressed the `WorkerLog` not being movable thusly:

![ekQwBPw 1](https://github.com/gui-cs/Terminal.Gui/assets/585482/725bbb99-474d-4b94-abc3-06d3a6167120)

tig commented on UnitTests/Dialogs/DialogTests.cs:
```diff
@@ -1331,4 +1331,37 @@ public void Zero_Buttons_Works ()
+        // This is instance and the user is responsible to set to null or not
+        // because in the Application it's all working as expected.
+        // Fortunately, this instance is not null so that we can obtain the value of its properties
+        Assert.NotNull (dlg);
```
Looking at this again... I'm confused.

This code literally says: "`dig` has been disposed, but it's ok to access it because the reference is not `null`".

This is not ok (and is actually the point I raised in filing #3338).

BDisp replied:
That's the user instance and not the Top and the Current instance. If the user instance is null how you can get a property value after the Current is disposed?

BDisp replied:
It's up to user disposing the instance and up to Application dispose the Current runstate.

BDisp replied:
I think you are looking for an issue where there is none.

tig reviewed (commented):
We're going in circles.

Am I wrong in asserting that it's not correct to access an object that has been disposed?

tig commented:
I've updated this PR to do the following:

- Moves the `Cancelled` property from `FileDialog` to `Dialog` so all dialogs can be checked for `Cancelled`
- Added a check to `Cancelled` get/set for `WasDisposed`
- Updated unit test to prove disposing Top in `Run/End` is incorrect (unit test current fails).

BDisp commented:
> Here's a new test that should pass (but doesn't):

You aren't testing this right.
Instead of `Assert.False (dlg.WasDisposed);`
Should be `Assert.True (dlg.WasDisposed);`

And in the `Canceled` property you must to

```cs
            if (!WasDisposed)
            {
                throw new ObjectDisposedException (GetType ().FullName);
            }
```

The PR that I submitted already do that way. You only asked why the `diag` isn't null. Because it's an instance but return the `WasDisposed` from the `runstate.Toplevel`. Can you explain why you think this is wrong?

tig commented:
> > Here's a new test that should pass (but doesn't):
> 
> You aren't testing this right. Instead of `Assert.False (dlg.WasDisposed);` Should be `Assert.True (dlg.WasDisposed);`
> 
> And in the `Canceled` property you must to
> 
> ```cs
>             if (!WasDisposed)
>             {
>                 throw new ObjectDisposedException (GetType ().FullName);
>             }
> ```
> 
> The PR that I submitted already do that way. You only asked why the `diag` isn't null. Because it's an instance but return the `WasDisposed` from the `runstate.Toplevel`. Can you explain why you think this is wrong?

Are you really telling me this code isn't correct?

```cs
        get
        {
#if DEBUG_IDISPOSABLE
            if (WasDisposed)
            {
                throw new ObjectDisposedException (GetType ().FullName);
            }
#endif
            return _canceled;
        }
```

BDisp commented:
> Are you really telling me this code isn't correct?
> 
> ```cs
>         get
>         {
> #if DEBUG_IDISPOSABLE
>             if (WasDisposed)
>             {
>                 throw new ObjectDisposedException (GetType ().FullName);
>             }
> #endif
>             return _canceled;
>         }
> ```

Yes because you are get the Canceled property after the runstate.Toplevel was disposed and it was really disposed. So only call it before.

tig commented:
Let me be more clear:

Are you saying the code I pasted above is the incorrect way to check if a disposed object is being accessed in general?

BDisp commented:
> Are you really telling me this code isn't correct?
> 
> ```cs
>         get
>         {
> #if DEBUG_IDISPOSABLE
>             if (WasDisposed)
>             {
>                 throw new ObjectDisposedException (GetType ().FullName);
>             }
> #endif
>             return _canceled;
>         }
> ```

Sorry, the code is correct but you are asserting after the dlg was really disposed. Assert it in the Dlg_Ready method.

BDisp commented:
This pass:

```cs
    [Fact]
    [AutoInitShutdown]
    public void Can_Access_Cancel_Property_After_Run ()
    {
        var top = Top;

        Dialog dlg = new ();

        dlg.Ready += Dlg_Ready;

        Run (dlg);

#if DEBUG_IDISPOSABLE
        Assert.True (dlg.WasDisposed);
        Assert.True (Top.WasDisposed);
        Assert.Equal (top, Top);
#endif

        Shutdown ();

        return;

        void Dlg_Ready (object sender, EventArgs e)
        {
            ((Dialog)sender).Canceled = true;
            Assert.True (dlg.Canceled);
            RequestStop ();
        }
    }

```

tig commented:
You are missing the point. 

'Run' is a synchronous method. Callers often need to get state of the toplevel they pass to run (eg Canceled,or in the case of FileDialog, the chosen file). The codebase is littered with code that does this. 

If, when 'Run' returns, the toplevel in question has been Disposed, IT IS AN INVALID OBJECT AND ACCESSING ITS PROPERTIES IS AN ERROR. IT IS ONLY LUCK IT WORKS IN MOST CASES TODAY (specifically because GC has not actually run).

BDisp commented:
> You are missing the point.
> 
> 'Run' is a synchronous method. Callers often need to get state of the toplevel they pass to run (eg Canceled,or in the case of FileDialog, the chosen file). The codebase is littered with code that does this.
> 
> If, when 'Run' returns, the toplevel in question has been Disposed, IT IS AN INVALID OBJECT AND ACCESSING ITS PROPERTIES IS AN ERROR. IT IS ONLY LUCK IT WORKS IN MOST CASES TODAY (specifically because GC has not actually run).

Ok I get the point but how you will guarantee that all derived View classes will dispose them after run?

BDisp commented:
You are right, must be the callers who have to dispose Toplevel. Must be possible to create some engineering that can control the last closed toplevel if it was disposed when calling Begin for a new toplevel or when the Shutdown was called. I'll also think better about this and try to help you.

tig commented:
Here's a thought:

Why have Init create and set Application.Top at all?

What if Run() (no parameters) acted just like Run<TopLevel>)?

Then, we remove the toplevel factory from Init. 

Then, we ensure Run<T>/End disposes the T that's created and Run(t) does not dispose the passed t object?

Any code that currently needs to modify Application.Top today will need to be modified to do

Application.Init();
Toplevel t = new();
...
Application.Run(t). 
t.Dispose();
Application.Shutdown();

Which is actually pretty clean.

BDisp commented:
> Here's a thought:
> 
> Why have Init create and set Application.Top at all?
> 
Unless we want to determine the real available terminal's area, but this can be achieved by the `Driver.Cols` and `Driver.Rows`. Also to it work as expected, e.g. to re-draw it need to being run. Thus my opinion is we don't need to create a toplevel on the `Init` method.

> What if Run() (no parameters) acted just like Run)?
> 
Didn't understanded this. I think you meaning by removing the Run() method and only use the Run(T). I agree and thus forces the user creating a toplevel before run.

> Then, we remove the toplevel factory from Init.
> 
Correct.

> Then, we ensure Run/End disposes the T that's created and Run(t) does not dispose the passed t object?
> 
Run/End only dispose the runstate created but not the passed T object, I agree.

> Any code that currently needs to modify Application.Top today will need to be modified to do
> 
> Application.Init();
> Toplevel t = new();
> ... Application.Run(t).
> t.Dispose();
> Application.Shutdown();
> 
> Which is actually pretty clean.

Yes, they needs to modified to forces the creating of the toplevel object and this is reasonable.

tig commented:
> > What if Run() (no parameters) acted just like Run)?
> 
> Didn't understanded this. I think you meaning by removing the Run() method and only use the Run(T). I agree and thus forces the user creating a toplevel before run.

What I mean is: there are currently 3 versions of `Application.Run`:

```cs
 public static void Run (Func<Exception, bool> errorHandler = null);
 public static void Run<T> (Func<Exception, bool> errorHandler = null, ConsoleDriver driver = null);
 public static void Run (Toplevel view, Func<Exception, bool> errorHandler = null);
```

In shorthand these are: `Run()`, `Run<T>`, and `Run(t)`.

What I'm proposing is the first one, `Run()`, currently does this:

```cs
public static void Run (Func<Exception, bool> errorHandler = null) { Run (Top, errorHandler); }
```

It would change to this:

```cs
public static void Run (Func<Exception, bool> errorHandler = null) { Run<Toplevel>(), errorHandler); }
```

BDisp commented:
It must be as the following, otherwise we it `throw new ArgumentException ($"{top.GetType ().Name} must be derived from TopLevel");`.

```cs
 public static void Run (Func<Exception, bool> errorHandler = null) { Run (new (), errorHandler); }
```

BDisp commented:
I submitted the PR https://github.com/tig/Terminal.Gui/pull/25, but it's still a WIP.

BDisp commented:
The `Run<T> Example` scenario isn't re-draw the `Dialog` correctly.

![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/7b092302-0f92-47bf-a666-f30495ef31ad)

tig commented:
@BDisp after merging your PR tests are failing. The failure is in 

![image](https://github.com/tig/Terminal.Gui/assets/585482/e97d55b5-060b-4437-91e5-8e53b9b6146f)

Any idea what's going on?

tig commented:
I just hacked in a fix.

BDisp commented:
> I just hacked in a fix.

Can you replace with this, please?

`_workerApp?.WriteLog ($"{(top.Toplevel is null ? ((Toplevel)sender).Data : top.Toplevel.Data)} activate.");`

I forgot to commit this before. Thanks.

tig commented:
@bdiisp please run BackgroundWorkercollection on your copy of this branch, show the Worker and then try to quit with ESC. Does it quit?

If you then use File.Quit does it throw?

BDisp commented:
> @bdiisp please run BackgroundWorkercollection on your copy of this branch, show the Worker and then try to quit with ESC. Does it quit?
> 
> If you then use File.Quit does it throw?

I sent a better solution for this because the intention is hiding it on pressing Esc. See https://github.com/tig/Terminal.Gui/pull/27.

tig commented:
Here's my thinking on `Scenario`. I'll either create a new issue/PR to address this or do it here, depending...

I over-engineered `Scenario` and tried to be both too helpful to devs and too cute. My intent was to make it super easy to create new scenarios. I thought it would be neat if Scenario developers only had to write a `Setup` override and the "default" logic in `Init` and `Run` would do magic for them. I also had UICatalog call `Application.Shutdown` when the Scenario exited instead of forcing Scenario authors to do this themselves.

This was a bad idea for these reasons:

1. New developers looking at scenarios as examples don't see what ones supposed to do to setup/teardown a Terminal.Gui app. They don't see the proper way: Call `Init`, and ensure it's bracketed by a `Shutdown`, Call `Run` and make sure the `Toplevel` used gets disposed. 
2. It led to un-needed complexity in 4 overrides: `Init`, `Setup`, `Run`, and `RequestStop`. This is confusing to devs and a survey of the existing scenarios shows this confusion.
3. It makes devs think that `Application.Top` can't be the actuall main view for the scenario, because `Init`, by default creates `Win` and adds it to `Top`. 

My proposal:

- Simplify `Scenario` to have only ONE virtrual metods: `Main`. Developers should treat this just like the standard C# console app `int Main (string [] args)` method.
- Add a `Application.Theme` property (a string) that sets the theme for the app. Do NOT reset this to `string.Empty` in `ResetState`. This way, when UI Catalog sets the Theme, it will be carried into any Scenario that's run. 
- Nuke the fancy UI Catalog concept of `ToplevelColorScheme` which was really just a test that is no longer really needed.
- Change all Scenarios to merge their `Init`, `Setup`, and `Run` overrides into `Main`. 

The Generic `Scenario.Main` will look like this:

```cs
    public override void Main ()
    {
        // Init
        Application.Init ();

        // Setup - Create a top-level application window and configure it.
        Window appWindow = new ()
        {
            Title = $"{Application.QuitKey} to Quit - Scenario: {GetName ()}",
        };

        var button = new Button { X = Pos.Center (), Y = Pos.Center (), Text = "Press me!" };
        button.Accept += (s, e) => MessageBox.ErrorQuery ("Error", "You pressed the button!", "Ok");
        appWindow.Add (button);

        // Run - Start the application.
        Application.Run (appWindow);
        appWindow.Dispose ();

        // Shutdown - Calling Application.Shutdown is required.
        Application.Shutdown ();
    }
```

BDisp commented:
Returning `Toplevel` only on the Run() and Run<T>() methods are fine. Void is more appropriate to the `Run(v)`.

tig commented:
> Returning `Toplevel` only on the Run() and Run_T_() methods are fine. Void is more appropriate to the `Run(v)`.
> 
> Why writing < and > are hiding?

Cool thing about `Run<T> and Run()` returning the toplevel is we can do this:

```cs
       Application.Top?.Dispose ();
```

tig commented:
I think this is ready to merge now. @BDisp you agree?

BDisp commented on UICatalog/Scenarios/BackgroundWorkerCollection.cs:
```diff
@@ -13,12 +14,25 @@ namespace UICatalog.Scenarios;
-    public override void Run () { Application.Run<OverlappedMain> (); }
+    public override void Init ()
+    {
+        Application.Run<OverlappedMain> ().Dispose ();
+        Application.Top?.Dispose ();
```
I don't know why you have `Dispose` twice and why you removed the `Application.Shutdown` which you want to force the users to use.

BDisp replied:
Should be:

```cs
        Application.Run<OverlappedMain> ().Dispose ();
        Application.Shutdown  ();
```

tig replied:
I misunderstood your code. I thought you were disposing Application.Top in addition to the view returned from Run because it may be a different view since SetCurrentOverlappedAsTop changed Top. 

Removing shutdown was an error on my part.

tig replied:
On your "you want to force users to call Shutdown" comment. 

Do you believe it's ok for apps to not call Application.Shutdown?

BDisp replied:
No, I mean mandatory. It's the right thing to do.

BDisp replied:
I think you're right also call Application.Top?.Dispose (); because it may not be the main. That's ensure not throwing the assertion.

tig replied:
> I think you're right also call `Application.Top?.Dispose ();` because it may not be the main. That's ensure not throwing the assertion.

When we get around to supporting any view in Run, and refactoring overlapped, we really need to make the concept of when Application.Top gets set more deterministic. It's scary that neither of us can articulate exactly why Top (and/or Current) has changed when Run returns here. 

But, for now, I'm not overly concerned because the only place it currently happens is in user code where someone is both using "TopLevelOverlapped" and background tasks/threading ... and this Scenario is pretty much it.

BDisp replied:
You are right. But this has a explanation. This can happen when we start a new RunWorker and close the scenario before it's be showing. I prevented them for not open if the `OverlappedTop` is null. I also fixed an issue where when the scenario is already closed, after start a new RunWorker and the async Post was is called after the `MainLoop` is null.
Maybe now the situation of the `Application.Top` isn't actually the `OverlappedTop`isn't now happen. I will remove the `Application.Top?.Dispose` and doing some tests to check it.

BDisp commented:
> I think this is ready to merge now. @BDisp you agree?

Yes, I agree. I only have the question above.

BDisp commented:
There is still a issue on the BackgroundWorkerCollection scenario.

- Open the scenario.
- Click on View menu and open the WorkApp
- Start two news Run Worker
- After they are open go to first that was open
- Press Esc to close it
- A blank window will be show
- Move the mouse over it
- System.NullReferenceException will be throw

![Captura de ecrã 2024-03-25 234626](https://github.com/gui-cs/Terminal.Gui/assets/13117724/5e1af468-dbe1-422d-baf6-96601d6071fb)

![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/5af93b6c-9f92-4b3c-98a7-632958af1502)

BDisp commented:
There is still a issue when the `OverlappedTop` is the `Current` after a `OverlappedChildren` was closed and there is still a `OverlappedChildren` visible which must be on the top of the toplevel's stack.

*To repo:
- Open the scenario.
- Click on View menu and open the WorkApp
- Start two news Run Worker
- After they are open go to first that was open
- Press Esc to close it
- Press Esc to close the current
- Press Esc to toggle the visibility of the WorkerApp to false
- The WorkerApp is still visible
- Press again Esc and now the WorkerApp is hidden

The WorkerApp must be hidden after pressing the first Esc

tig commented:
I found another bug in BWC:

- Ctrl-R to start a new worker
- ENTER to accept
- ListView crashes

The reason is when you create the staging UI on Ctrl-R you create the listview but don't populate it. By default it has focus, so ENTER goes to the listview. But there's no items so ListView crashes. 

The fix, which I've put in already, is to create the listview `Enabled = false` and in the constructor where you set the listview Source, do `_listView.Enabled = true`.

tig commented:
@BDisp - Ok if I merge this to v2_develop now? further bugs in BWC Scenario can be fixed in another PR?

BDisp commented:
> @BDisp - Ok if I merge this to v2_develop now? further bugs in BWC Scenario can be fixed in another PR?

Yes.

BDisp commented:
> I found another bug in BWC:
> 
> * Ctrl-R to start a new worker
> * ENTER to accept
> * ListView crashes
> 
> The reason is when you create the staging UI on Ctrl-R you create the listview but don't populate it. By default it has focus, so ENTER goes to the listview. But there's no items so ListView crashes.
> 
> The fix, which I've put in already, is to create the listview `Enabled = false` and in the constructor where you set the listview Source, do `_listView.Enabled = true`.

This started given the error because you enabled the `ListView` to start the `RunWorker`. I only enabled it after the work is finished to show the items. The ListView had a bug really but why you enabled it before on the window that will start the work?

tig commented:
> @tig you didn't change as I did in my PR https://github.com/tig/Terminal.Gui/pull/29/files. Sometimes is preferable I merge and resolving the conflicts with your branch than you manually change them.
> 
> ![image](https://private-user-images.githubusercontent.com/13117724/316972301-9096d2ab-e722-42a5-9c08-316dfd6244b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NzUyODksIm5iZiI6MTcxMTQ3NDk4OSwicGF0aCI6Ii8xMzExNzcyNC8zMTY5NzIzMDEtOTA5NmQyYWItZTcyMi00MmE1LTljMDgtMzE2ZGZkNjI0NGI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDE3NDMwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiOWY1YTAxMzM2ZWUyYmM1MTJlMDdkZGI0OWI5NmI2NGNhY2RkZjA1MDlkMTQ2MWVkNzI2MDY4ODQ0OTE5YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.48nV_OFyQfJCyhLuLhLvbYnqQGVgk65fhbxvH66Lrxg)

Yes, I missed removing the Dispose call on line 488.

However, your code is incorrect. You can't dispose an object and then reference. It needs to be like this:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/e4c1b779-79b1-4845-aa50-05ea1571d428)

BDisp commented:
> However, your code is incorrect. You can't dispose an object and then reference. It needs to be like this:

It's correct because bellow is another object and not the same. See it again.

tig commented:
> > However, your code is incorrect. You can't dispose an object and then reference. It needs to be like this:
> 
> It's correct because bellow is another object and not the same. See it again.

Nope. You're missing it. In no universe is this code correct:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/c60e04d0-f6dd-480f-97e9-0c4bd1ed42b7)

BDisp commented:
Edit:
Ah I see now, sorry. I didn't see the image before.

Here my suggestion:

```cs
            Application.Run (stagingUI);

            if (stagingUI.Staging != null && stagingUI.Staging.StartStaging != null)
            {
                stagingUI.Dispose ();

                staging = new Staging (stagingUI.Staging.StartStaging);
                WriteLog ($"Worker is started at {staging.StartStaging}.{staging.StartStaging:fff}");

                if (_stagingWorkers == null)
                {
                    _stagingWorkers = new Dictionary<Staging, BackgroundWorker> ();
                }

                _stagingWorkers.Add (staging, worker);
                worker.RunWorkerAsync ();
            }
            else
            {
                stagingUI.Dispose ();
            }

```

tig commented:
":

> Edit: Ah I see now, sorry. I didn't see the image before.
> 
> Here my suggestion:
> 
> ```cs
>             Application.Run (stagingUI);
> 
>             if (stagingUI.Staging != null && stagingUI.Staging.StartStaging != null)
>             {
>                 stagingUI.Dispose ();
> 
>                 staging = new Staging (stagingUI.Staging.StartStaging);
>                 WriteLog ($"Worker is started at {staging.StartStaging}.{staging.StartStaging:fff}");
> 
>                 if (_stagingWorkers == null)
>                 {
>                     _stagingWorkers = new Dictionary<Staging, BackgroundWorker> ();
>                 }
> 
>                 _stagingWorkers.Add (staging, worker);
>                 worker.RunWorkerAsync ();
>             }
>             else
>             {
>                 stagingUI.Dispose ();
>             }
> ```

Where does the `staging = new ...` instance get disposed?

BDisp commented on UICatalog/Scenarios/BackgroundWorkerCollection.cs:
```diff
@@ -90,29 +105,22 @@ public OverlappedMain ()
-                _canOpenWorkerApp = true;
+                Application.Run (_workerApp);
 
                 return;
```
This return isn't necessary.

tig commented:
This is STILL incorrect code:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/22b716d2-1945-4263-9993-987ac21d55d8)

BDisp commented:
> Where does the `staging = new ...` instance get disposed?
Here.

![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/0ab1ac2f-472c-428d-a985-fbc6e27e89b9)

BDisp commented:
> This is STILL incorrect code:
> 
> ![image](https://private-user-images.githubusercontent.com/585482/316987357-22b716d2-1945-4263-9993-987ac21d55d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NzY2MjYsIm5iZiI6MTcxMTQ3NjMyNiwicGF0aCI6Ii81ODU0ODIvMzE2OTg3MzU3LTIyYjcxNmQyLTE5NDUtNDI2My05OTkzLTk4N2FjMjFkNTVkOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNlQxODA1MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mODA4NWE3NjBjNmEzZDcwMzAwOTMxOWViZDJhYzQ4YjA4MGMxMmIxZDQ2MjA5ZGYzNDM2MzAyM2RlM2I1Y2JlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.0iE7YrQ40eGFeO6c7RIKKglBdrx1O3SWIfa1NI9qFlM)

You are right. Put it before `worker.RunWorkerAsync ();`

tig commented:
> > Where does the `staging = new ...` instance get disposed?
> > Here.
> 
> ![image](https://private-user-images.githubusercontent.com/13117724/316987594-0ab1ac2f-472c-428d-a985-fbc6e27e89b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NzY2ODAsIm5iZiI6MTcxMTQ3NjM4MCwicGF0aCI6Ii8xMzExNzcyNC8zMTY5ODc1OTQtMGFiMWFjMmYtNDcyYy00MjhkLWE5ODUtZmJjNmUyN2U4OWI5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDE4MDYyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3MzMwMjFkY2RlNTk0NWZhYTYxZjg4ZWJiZWY3MzAwZWFhNzQ4ODhmMTE2NDQ0YTIwZjk3YTNmNTJhOWJjYTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.G-G3KbiEyiVIXl1c567IZvxNvGQQ4rdQPI3p3xb-_QU)

Given this, then this seems correct:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/0f29f30b-74bc-4031-9b36-ee34fa76e788)

Agree?

BDisp commented on UICatalog/Scenarios/BackgroundWorkerCollection.cs:
```diff
@@ -493,6 +492,7 @@ public void RunWorker ()
             if (stagingUI.Staging != null && stagingUI.Staging.StartStaging != null)
             {
                 staging = new Staging (stagingUI.Staging.StartStaging);
+                stagingUI.Dispose ();
```
Not yet in the right place. 😃

BDisp commented:
That's now right. The scenario had a great refactoring. It's much more intuitive.

dodexahedron commented:
The ultimate solution to this (and a bunch of other things, actually), which I strongly encourage for post-V2, is to use the built-in DI container infrastructure in .net and just inject things into it when and as needed.

That also vastly simplifies things like testing, logging, configuration, and background tasks/services, each in appropriate scopes, plus being a mature, trusted, and standard core of modern .net applications, and also makes extending it even easier.

It's actually not that much effort to port existing code to, and could be done fairly easily for v2 (usually you can so it with almost no changes, in fact, and just get all the goodies for free), but I think we have enough on our plates for v2 already and gotta draw the line _somewhere._

Ok PC says 98% so maybe I can log in shortly, now. 😅
